### PR TITLE
feat(sipi): Improve error message if XSL file not found

### LINKED ISSUE #1580: Improve error message with more details about the failure
**Issue description**
The full text search has been tested on Knora-app after checking out Knora-api develop branch with the latest commit. The data has been reloaded with `make init-db-test`. The search failed with some terms like "suisse" and "france". An error 404 has been thrown (see screen shot below).

**What must have happened** 
I forgot to move the server file of BEOL into the images folder of BEOL.
In the BEOL mappings, a default XSL transformation is specified. This is a file hosted by Sipi which Knora requested but Sipi could not serve because it did not find it.

**What we would like**
A better explained error message would have been helpful to understand the situation.
Knora makes an internal request to Sipi for the XSLT file. If this request fails, Knora's error message should say that it tried to get this file from Sipi but got an unsuccessful response.

**Screen shots**
Error in the console of Chrome:
Screenshot 2020-01-23 at 15 14 37

### PR DESCRIPTION
**Pull request by benjamingeer**
This PR improves the error returned if Knora requests a nonexistent XSL transformation file from Sipi.

Resolves #1580.

### CODE REVIEW

**Comment by benjamingeer** [comment]
@flavens Could you check if this branch gives you a better error message?

**Comment by flavens** [comment]
> @flavens Could you check if this branch gives you a better error message?

I get back this message:
Screenshot 2020-02-05 at 17 03 37

I think it is better, however, I would have search for a long time it was the BEOL server file at the wrong place! I still need more experience with errors to get a better instinct :)

**Comment by benjamingeer** [comment]
@flavens Hmm, that's not what it's supposed to say. Can you give me a simple way to make this error happen?

**Comment by benjamingeer** [comment]
It's supposed to say something like "Unable to get XSL transformation file XYZ from Sipi"...

**Comment by flavens** [comment]
> @flavens Hmm, that's not what it's supposed to say. Can you give me a simple way to make this error happen?

You can checkout the master branch of Knora-app.

Run on your terminal: `yarn install --prod=false` + `ng s` (you have to install yarn and [angular-cli](https://www.npmjs.com/package/@angular/cli#installation))

Then go to http://0.0.0.0:4200/system/users after log in (you can use the root user profile) and try to **Add a user as system admin**:

Screenshot 2020-02-06 at 08 26 37

**Comment by benjamingeer** [comment]
```
% yarn install --prod=false + ng s
yarn install v1.22.0
error `install` has been replaced with `add` to add new dependencies. Run "yarn add + ng s" instead.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

**Comment by benjamingeer** [comment]
```
% yarn add --prod=false + ng s
yarn add v1.22.0
[1/4] 🔍  Resolving packages...
error An unexpected error occurred: "https://registry.yarnpkg.com/+: Not found".
info If you think this is a bug, please open a bug report with the information provided in "/Users/benjamingeer/git/knora-app/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
```

**Comment by flavens** [comment]
> ```
> % yarn add --prod=false + ng s
> yarn add v1.22.0
> [1/4] 🔍  Resolving packages...
> error An unexpected error occurred: "https://registry.yarnpkg.com/+: Not found".
> info If you think this is a bug, please open a bug report with the information provided in "/Users/benjamingeer/git/knora-app/yarn-error.log".
> info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
> ```

you run `yarn **install**` not `yarn add` first (make sure yarn is installed) and then,` ng s`(with angular-cli installed on your computer)

**Comment by benjamingeer** [comment]
> you run yarn **install** not yarn add first

I did that and got an error message that told me to run `yarn install` instead, as shown in my comment above:

https://github.com/dasch-swiss/knora-api/pull/1590#issuecomment-582786436

The error message says "error `install` has been replaced with `add` to add new dependencies. Run "yarn add + ng s" instead."

But `yarn add` doesn't work either, it gives a different error message.

**Comment by musicEnfanthen** [comment]
Can confirm that the whole command `yarn add --prod=false + ng s` is not working. You would have to run it sequentially:

`yarn add --prod=false `

[wait for yarn to complete installation]

`ng s` [optional with `ng s -o`, then the app opens directly in the browser]

**Comment by benjamingeer** [comment]
@musicEnfanthen Thanks, I just figured out the same thing at the same time. :)

**Comment by benjamingeer** [comment]
@flavens I did what you described, and I can see in the console that the request is empty (`{}`). That's why you're getting the error "The request content was malformed: No data sent in API request." Maybe that's because of a bug in `knora-app`.

In any case, I don't think that can have anything to do with the issue about an XSL file that doesn't exist. The request to create an admin user uses the admin API, which doesn't do anything with XSL templates. XSL templates are used to transform text markup in a resource.

**Comment by flavens** [comment]
@benjamingeer ah! an idea: I spotted this error before the release of knora-app. In the master branch, we have the newest knora-api-js-lib. To reproduce the error, you could try with knora-api-js-lib v0.1.1, knora-api v11.0.0 and place the BEOL server file in another folder.

**Comment by benjamingeer** [comment]
@flavens OK, I've actually just added an e2e test for this, which confirms that if you request a resource with an XSL transformation, and the XSL file doesn't exist, you get an HTTP 500 error with the message "Unable to get XSL transformation file".

I don't know how to check this manually using the BEOL data, but I guess you would have to request a BEOL resource. @tobiasschweizer can you help?

File: webapi/src/main/scala/org/knora/webapi/store/iiif/SipiConnector.scala [hunk 1]
```diff
@@ -321,9 +321,14 @@ class SipiConnector extends Actor with ActorLogging {
         // ask Sipi to return the XSL transformation
         val request = new HttpGet(xsltFileUrl)
 
-        for {
+        val sipiResponseTry: Try[SipiGetTextFileResponse] = for {
             responseStr <- doSipiRequest(request)
         } yield SipiGetTextFileResponse(responseStr)
+
+        sipiResponseTry.recover {
+            case badRequestException: BadRequestException => throw SipiException(s"Unable to get XSL transformation file $xsltFileUrl from Sipi: ${badRequestException.message}")
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/store/iiif/SipiConnector.scala]: `sipiGetXsltTransformationRequestV2` is called in two cases: for XSLT used by `StandoffResponderV2` to turn XML into HTML and for XSLT used the TEI transformer in `ResourcesResponderV2`. So I guess testing it with the TEI route is fine.

However, `SipiGetTextFileRequest` is as the message sent to `SipiConnector` that then internally calls `sipiGetXsltTransformationRequestV2` seems strange to me. Shouldn't either the message be more specific or the method more generic?

Comment (tobiasschweizer) [reply webapi/src/main/scala/org/knora/webapi/store/iiif/SipiConnector.scala]: In the future, Knora might require other text files from Sipi than XSL transformations and then the error message thrown by `sipiGetXsltTransformationRequestV2 ` might be misleading.

Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/store/iiif/SipiConnector.scala]: Quite true, I'll fix that.

Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/store/iiif/SipiConnector.scala]: Done in 080dd44.

**Comment by benjamingeer** [comment]
@tobiasschweizer @flavens Is this better now?

**Comment by benjamingeer** [comment]
> Is there a way to tell the client why Knora tried to get a text file from Sipi?

What would you like it to say?

**Comment by tobiasschweizer** [comment]
> What would you like it to say?

It would be great if the message would sell who sent the message to `SipiResponder`, e.g. the `StandoffResponderV2`.

**Comment by benjamingeer** [comment]
You'd like it to say, "Requested by StandoffResponderV2"?

**Comment by tobiasschweizer** [comment]
> You'd like it to say, "Requested by StandoffResponderV2"?

Yes, but only if the message was sent by `StandoffResponder`. Can you find out about this in `SipoResponder`?

**Comment by benjamingeer** [comment]
> Can you find out about this in SipoResponder?

I could add it to the message.

**Comment by tobiasschweizer** [comment]
> I could add it to the message.

That would great. It is like a stack trace.

**Comment by benjamingeer** [comment]
OK, now the error message is "Unable to get file http://0.0.0.0:1024/0801/missing.xsl from Sipi as requested by org.knora.webapi.responders.v2.StandoffResponderV2: Sipi responded with HTTP status code 404: Not Found".

**Comment by benjamingeer** [comment]
@tobiasschweizer @flavens Thanks to both of you for your help with this.

**Pull request merged**